File: backend/app/routers/classify.py
# app/routers/classify.py
from fastapi import APIRouter, BackgroundTasks, HTTPException
from app.models.classify import ManualLabelRequest, SingleAnnotateRequest, NaiveLabelRequest, ClusteringRequest
from app.services.classify_service import create_classify_job, run_manual_annotation_job, run_naive_job, run_clustering_job
from app.db.supabase_client import supabase

router = APIRouter(prefix="/datasets", tags=["classification"])

@router.post("/{dataset_id}/label/manual")
def label_manual(dataset_id: str, req: ManualLabelRequest, background_tasks: BackgroundTasks):
    # validate dataset exists and belongs to session
    res = supabase.table("datasets").select("*").eq("dataset_id", dataset_id).eq("session_id", req.session_id).execute()
    if not res.data:
        raise HTTPException(status_code=404, detail="Dataset not found")

    job_id = create_classify_job(dataset_id, req.session_id, "manual")
    # schedule
    background_tasks.add_task(run_manual_annotation_job, job_id, dataset_id, req.session_id, [a.dict() for a in req.annotations])
    return {"job_id": job_id}

@router.post("/{dataset_id}/label/manual/row")
def label_single_row(dataset_id: str, req: SingleAnnotateRequest, background_tasks: BackgroundTasks):
    # implement by creating a tiny job that annotates single row
    job_id = create_classify_job(dataset_id, req.session_id, "manual_single")
    background_tasks.add_task(run_manual_annotation_job, job_id, dataset_id, req.session_id, [{"row_index": req.row_index, "label": req.label}])
    return {"job_id": job_id}

@router.post("/{dataset_id}/label/naive")
def label_naive(dataset_id: str, req: NaiveLabelRequest, background_tasks: BackgroundTasks):
    res = supabase.table("datasets").select("*").eq("dataset_id", dataset_id).eq("session_id", req.session_id).execute()
    if not res.data:
        raise HTTPException(status_code=404, detail="Dataset not found")

    job_id = create_classify_job(dataset_id, req.session_id, "naive")
    background_tasks.add_task(run_naive_job, job_id, dataset_id, req.session_id, req.keyword_map, req.use_default_keywords)
    return {"job_id": job_id}

@router.post("/{dataset_id}/label/clustering")
def label_clustering(dataset_id: str, req: ClusteringRequest, background_tasks: BackgroundTasks):
    res = supabase.table("datasets").select("*").eq("dataset_id", dataset_id).eq("session_id", req.session_id).execute()
    if not res.data:
        raise HTTPException(status_code=404, detail="Dataset not found")

    job_id = create_classify_job(dataset_id, req.session_id, req.algorithm)
    background_tasks.add_task(run_clustering_job, job_id, dataset_id, req.session_id, req.algorithm, req.hyperparameters)
    return {"job_id": job_id}

# job status fetcher
@router.get("/classify_jobs/{job_id}")
def get_job(job_id: str):
    res = supabase.table("classify_jobs").select("*").eq("job_id", job_id).execute()
    if not res.data:
        raise HTTPException(status_code=404, detail="Job not found")
    return res.data[0]

File: frontend/src/api/classify.ts
// src/api/classify.ts
import axios from "axios";

const API_BASE = "https://machine-learning-tweet-sentiment.onrender.com/datasets";

// -------------------------
// Types from FastAPI models
// -------------------------
export interface ManualAnnotation {
  row_index: number;
  label: string;
}

export interface ManualLabelRequest {
  session_id: string;
  annotations: ManualAnnotation[];
}

export interface SingleAnnotateRequest {
  session_id: string;
  row_index: number;
  label: string;
}

export interface NaiveLabelRequest {
  session_id: string;
  keyword_map?: Record<string, string[]>; // { positive: ["good"], negative: ["bad"] }
  use_default_keywords: boolean;
}

export interface ClusteringRequest {
  session_id: string;
  algorithm: string;
  hyperparameters: Record<string, any>;
}

// -------------------------------
// Start Manual (Batch) Label Job
// -------------------------------
export const labelManual = async (
  datasetId: string,
  req: ManualLabelRequest
) => {
  const res = await axios.post(`${API_BASE}/${datasetId}/label/manual`, req);
  return res.data as { job_id: string };
};

// -------------------------------
// Manual Label - Single Row
// -------------------------------
export const labelSingleRow = async (
  datasetId: string,
  req: SingleAnnotateRequest
) => {
  const res = await axios.post(`${API_BASE}/${datasetId}/label/manual/row`, req);
  return res.data as { job_id: string };
};

// -------------------------------
// Naive Automatic Labeling
// -------------------------------
export const labelNaive = async (
  datasetId: string,
  req: NaiveLabelRequest
) => {
  const res = await axios.post(`${API_BASE}/${datasetId}/label/naive`, req);
  return res.data as { job_id: string };
};

// -------------------------------
// Clustering Based Labeling
// -------------------------------
export const labelClustering = async (
  datasetId: string,
  req: ClusteringRequest
) => {
  const res = await axios.post(`${API_BASE}/${datasetId}/label/clustering`, req);
  return res.data as { job_id: string };
};

// -------------------------------
// Fetch Classification Job Status
// -------------------------------
export const getClassifyJob = async (jobId: string) => {
  const res = await axios.get(
    `${API_BASE}/classify_jobs/${jobId}`
  );

  return res.data as {
    job_id: string;
    dataset_id: string;
    session_id: string;
    status: "pending" | "running" | "completed" | "error";
    created_at?: string;
    finished_at?: string;
    logs?: string;
  };
};

// -------------------------------
// Poll Until Classification Done
// -------------------------------
export const waitForClassificationCompletion = async (
  jobId: string,
  onUpdate?: (status: string) => void,
  intervalMs: number = 1500
): Promise<any> => {
  return new Promise((resolve, reject) => {
    const interval = setInterval(async () => {
      try {
        const job = await getClassifyJob(jobId);

        if (onUpdate) onUpdate(job.status);

        if (job.status === "completed") {
          clearInterval(interval);
          resolve(job);
        }

        if (job.status === "error") {
          clearInterval(interval);
          reject(new Error(job.logs || "Classification failed."));
        }
      } catch (err) {
        clearInterval(interval);
        reject(err);
      }
    }, intervalMs);
  });
};